drivers/watchdog/wdt_nxp_ewm.c

```diff
@@ -0,0 +1,193 @@
+
+#include <zephyr/drivers/watchdog.h>
+#include <zephyr/irq.h>
+#include <fsl_clock.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <zephyr/logging/log.h>
+
+#define NXP_EWM_FEED_MAGIC_NUMBER 0x2CB4
+#define NXP_EWM_MAX_TIMEOUT_WINDOW 0xFE
+
+struct nxp_ewm_config {
+	EWM_Type *base;
+	void (*irq_config_func)(const struct device *dev);
+	bool is_input_enabled;
+	bool is_input_active_high;
+	uint8_t clk_divider;
+};
+
+struct nxp_ewm_data {
+	struct wdt_timeout_cfg timeout_cfg;
+	bool is_watchdog_setup;
+};
+
+static int nxp_ewm_setup(const struct device *dev, uint8_t options)
+{
+	const struct nxp_ewm_config *config = dev->config;
+	struct nxp_ewm_data *data = dev->data;
+	EWM_Type *base = config->base;
+
+	if (data->is_watchdog_setup) {
+		/* Watchdog cannot be re-configured after enabled. */
+		return -EBUSY;
+	}
+
+	if (options) {
+		/* Unable to halt counter during debugging */
+		return -ENOTSUP;
+	}
+
+	data->is_watchdog_setup = true;
+	base->CMPL = EWM_CMPL_COMPAREL(data->timeout_cfg.window.min);
+	base->CMPH = EWM_CMPH_COMPAREH(data->timeout_cfg.window.max);
+
+	/*
+	 * base->CTRL should be the last thing touched due to
+	 * the small watchdog window time.
+	 * After this write, only the INTEN bit is writable until reset.
+	 *
+	 * EWM_CTRL_INTEN enables the interrupt signal
+	 * EWM_CTRL_EWMEN enables the watchdog.
+	 */
+	base->CTRL |= EWM_CTRL_INEN(config->is_input_enabled)	|
+		EWM_CTRL_ASSIN(config->is_input_active_high)	|
+		EWM_CTRL_INTEN(1) | EWM_CTRL_EWMEN(1);
+
+	return 0;
+}
+
+static int nxp_ewm_disable(const struct device *dev)
+{
+	struct nxp_ewm_data *data = dev->data;
+
+	if (!data->is_watchdog_setup) {
+		return -EFAULT;
+	}
+
+	return -EPERM;
+}
+
+static int nxp_ewm_install_timeout(const struct device *dev,
+				     const struct wdt_timeout_cfg *cfg)
+{
+	struct nxp_ewm_data *data = dev->data;
+
+	if (cfg->flags) {
+		return -ENOTSUP;
+	}
+
+
+	if (data->is_watchdog_setup) {
+		return -ENOMEM;
+	}
+
+	if (cfg && cfg->window.max <= NXP_EWM_MAX_TIMEOUT_WINDOW &&
+		cfg->window.min <= cfg->window.max &&
+		cfg->window.max > 0 &&
+		cfg->window.min >= 0) {
+		data->timeout_cfg.window = cfg->window;
+	} else {
+		return -EINVAL;
+	}
+
+#if defined(CONFIG_WDT_MULTISTAGE)
+	if (cfg->next) {
+		return -EINVAL;
+	}
+#else
+
+	if (cfg->callback) {
+		data->timeout_cfg.callback = cfg->callback;
+	}
+
+	return 0;
+}
+
+static int nxp_ewm_feed(const struct device *dev, int channel_id)
+{
+	ARG_UNUSED(channel_id);
+	const struct nxp_ewm_config *config = dev->config;
+	EWM_Type *base = config->base;
+	unsigned int key = irq_lock();
+
+	base->SERV = EWM_SERV_SERVICE(NXP_EWM_FEED_MAGIC_NUMBER);
+	base->SERV = EWM_SERV_SERVICE((uint8_t)(NXP_EWM_FEED_MAGIC_NUMBER >> 8));
+	irq_unlock(key);
+
+	return 0;
+}
+
+static void nxp_ewm_isr(const struct device *dev)
+{
+	const struct nxp_ewm_config *config = dev->config;
+	struct nxp_ewm_data *data = dev->data;
+	EWM_Type *base = config->base;
+	base->CTRL &= (~EWM_CTRL_INTEN_MASK);
```

Comment:
im surprised if the check compliance doesnt complain, declarations should be separated from following code by newline

Reply:
Updated